Initial Kokkos implementation of NemoLite2D

sergisiso opened the pull request:
The initial implementation has the kokkos parallel dispatch and the Execution Policies (for the host), but it doesn't include targeting acceleration devices or Views (which are necessary for performance portability).

For the host it matches the C++ OpenMP (which it uses under-the-hood), so it doesn't introduce any overhead at least for the simple implementation.
```
Implementation          Compiler        Arch    ua checksum     uv checksum     time/step
psykal_serial:          gcc             cpu     0.41022150E-01  0.50252378E+01  0.15323E+00
psykal_cpp:             gcc             cpu     0.41022150E-01  0.50252378E+01  0.11709E+00
psykal_omp:             gcc             cpu     0.41022150E-01  0.50252378E+01  0.35053E-01
psykal_cpp_omp:         gcc             cpu     0.41022150E-01  0.50252378E+01  0.29249E-01
psykal_kokkos:          gcc             cpu     0.41022150E-01  0.50252378E+01  0.29679E-01
```

sergisiso commented:
@arporter @rupertford The initial (host only / no data layout abstraction) kokkos version is ready for review. It is conceptually quite similar to the C++ OpenMP version with only a different syntax. When writing the README of this implementation I realized I didn't add one for the `psykal_cpp` , so it is also part of this PR.

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/README.md:
```diff
@@ -0,0 +1,49 @@
+NEMOLite2D that has had its PSy layers manually implemented in C++.
+
+The Fortran(NemoLite2d) to C++(PSy-layer) interoperability is done
+with a `iso_c_binding` interface of the function `c_invoke_time_step`
```
s/a/an/ and s/of the/for the/

sergisiso replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/Makefile:
```diff
@@ -0,0 +1,69 @@
+# Makefile for manual C++ kokkos version of PSyKAl'd NEMOLite2D.
+#
+# This Makefile picks up the compiler to use any flags from
```
s/to use any/to use plus any/

sergisiso replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/README.md:
```diff
@@ -0,0 +1,63 @@
+
+## Implementation limitations ##
+
+Kokkos has several abstraction to control where (execution spaces),
```
s/abstraction/abstractions/ and s/,/and/

sergisiso replied:
Done

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:
```diff
@@ -0,0 +1,186 @@
+    );
+
+    // Time update kernel (internal domain u points)
+    Kokkos::parallel_for("continuity",
```
s/continuity/next_sshu/ I think?

sergisiso replied:
Ups my copy-pasting shows here :)

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:
```diff
@@ -0,0 +1,186 @@
+    );
+
+    // Time update kernel (internal domain v points)
+    Kokkos::parallel_for("continuity",
```
s/continuity/next_sshv/ I think?

sergisiso replied:
Fixed

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:
```diff
@@ -0,0 +1,186 @@
+    );
+
+    // Copy 'next' fields to 'current' fields (whole domain)
+    Kokkos::parallel_for("bc_solid_v",
```
s/bc_solid_v/copy_fields/ or similar?

sergisiso replied:
Fixed

sergisiso commented:
@arporter I fixed the typos and corrected the momentum loop boundaries of C++ and kokkos like in the upcoming PR #37 . This is ready for the next review.

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:
```diff
@@ -84,10 +84,10 @@ extern "C" void c_invoke_time_step(
     );
 
-    // Momentum_u kernel (internal domain)
+    // Momentum_u kernel (internal domain but top x)
```
Apologies if I'm being picky but I don't understand what 'but top x' means. Perhaps you want "internal u points" and similar for v below?

sergisiso replied:
Yes, that sounds better, changed.

sergisiso commented:
@arporter Fixed the remaining issue, ready for next review
